File: app-android/src/main/java/com/ustadmobile/port/android/sync/UmAppDatabaseSyncWorker.java
package com.ustadmobile.port.android.sync;

import android.content.Context;
import android.support.annotation.NonNull;

import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.impl.UMLog;
import com.ustadmobile.core.impl.UmAccountManager;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.lib.db.entities.UmAccount;
import com.ustadmobile.port.android.netwokmanager.UmAppDatabaseSyncService;

import java.util.concurrent.TimeUnit;

import androidx.work.Constraints;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

public class UmAppDatabaseSyncWorker extends Worker {

    public static final String TAG = "UmAppDbSync";

    public UmAppDatabaseSyncWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    public static void queueSyncWorker(long delay, TimeUnit timeUnit) {
        Constraints workConstraint = new Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .build();
        OneTimeWorkRequest request = new OneTimeWorkRequest.Builder(UmAppDatabaseSyncWorker.class)
                .setInitialDelay(delay, timeUnit)
                .addTag(TAG)
                .setConstraints(workConstraint)
                .build();
        WorkManager.getInstance().enqueue(request);
    }


    @NonNull
    @Override
    public Result doWork() {
        UmAccount activeAccount = UmAccountManager.getActiveAccount(getApplicationContext());

        UmAppDatabase umAppDb = UmAppDatabase.getInstance(getApplicationContext());
        try {
            umAppDb.syncWith(
                    UmAccountManager.getRepositoryForActiveAccount(getApplicationContext()),
                    activeAccount != null ? activeAccount.getPersonUid() : 0, 100, 100);
            UstadMobileSystemImpl.l(UMLog.INFO, 100, "database syncWith repo ran");
        }catch(Exception e) {
            UstadMobileSystemImpl.l(UMLog.WARN, 101, "Exception running syncWith :" +
                    e.getMessage());
        }

        if(!isStopped()) {
             boolean appRecentlyActive = UmAppDatabaseSyncService.isInForeground()
                    || System.currentTimeMillis() - UmAppDatabaseSyncService.getLastForegroundTime()
                    < UmAppDatabaseSyncService.SYNC_AFTER_BACKGROUND_LAG;

             if(appRecentlyActive ||
                     umAppDb.countPendingLocalChanges(UmAccountManager.getActivePersonUid(
                             getApplicationContext()), umAppDb.getDeviceBits()) > 0) {
                 queueSyncWorker(appRecentlyActive ? 1 : 15, TimeUnit.MINUTES);
             }
        }


        return Result.success();
    }
}

File: app-android/src/main/java/com/ustadmobile/port/android/sync/UmAppDatabaseSyncWorker.kt
package com.ustadmobile.port.android.sync

import android.content.Context
import androidx.work.*
import com.ustadmobile.core.account.Endpoint
import com.ustadmobile.core.account.EndpointScope
import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.db.UmAppDatabase.Companion.TAG_REPO
import com.ustadmobile.core.impl.UMLog
import com.ustadmobile.door.DoorDatabaseSyncRepository
import com.ustadmobile.port.android.netwokmanager.UmAppDatabaseSyncService
import kotlinx.coroutines.runBlocking
import org.kodein.di.DI
import org.kodein.di.android.di
import org.kodein.di.instance
import org.kodein.di.on
import java.util.concurrent.TimeUnit

class UmAppDatabaseSyncWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {

    override fun doWork(): Result {
        val di: DI by di(applicationContext)

        //should not be needed anymore
//        EndpointScope.Default.activeEndpointUrls.forEach { endpointUrl ->
//            try {
//                val clientRepo: UmAppDatabase by di.on(Endpoint(endpointUrl)).instance(tag = TAG_REPO)
//                runBlocking {
//                    (clientRepo as? DoorDatabaseSyncRepository)?.sync(null)
//                }
//            } catch (e: Exception) {
//                UMLog.l(UMLog.WARN, 101, "Exception running syncWith for $endpointUrl:" + e.message)
//            }
//        }


        if (!isStopped) {
            val appRecentlyActive = UmAppDatabaseSyncService.isInForeground ||
                    System.currentTimeMillis() - UmAppDatabaseSyncService.lastForegroundTime <
                    UmAppDatabaseSyncService.SYNC_AFTER_BACKGROUND_LAG
            if (appRecentlyActive) {
                queueSyncWorker((if (appRecentlyActive) 1 else 15).toLong(), TimeUnit.MINUTES)
            }
        }


        return Result.success()
    }


    companion object {

        val TAG = "UmAppDbSync"

        fun queueSyncWorker(delay: Long, timeUnit: TimeUnit) {
            val workConstraint = Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()
            val request = OneTimeWorkRequest.Builder(UmAppDatabaseSyncWorker::class.java)
                    .setInitialDelay(delay, timeUnit)
                    .addTag(TAG)
                    .setConstraints(workConstraint)
                    .build()
            WorkManager.getInstance().enqueue(request)
        }

        /**
         * It will run the new OneTimeWorkRequests only if
         *             there is no pending work labelled with the TAG
         */
        fun queueSyncWorkerWithPolicy(delay: Long, timeUnit: TimeUnit,
                                      policy: ExistingWorkPolicy) {
            val workConstraint = Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()
            val request = OneTimeWorkRequest.Builder(UmAppDatabaseSyncWorker::class.java)
                    .setInitialDelay(delay, timeUnit)
                    .addTag(TAG)
                    .setConstraints(workConstraint)
                    .build()
            WorkManager.getInstance().enqueueUniqueWork(TAG, policy, request)
        }
    }
}
